Show counts including all variants in order cycle exchanges

Matt-Yorkley opened the pull request:
#### What? Why?

Closes #4594 

<!-- Explain why this change is needed and the solution you propose.
Provide context for others to understand it. -->

Adapts order cycle exchange panels to show full count of all variants.

#### What should we test?
<!-- List which features should be tested and how. -->

Editing an order cycle: exchanges show full count of all variants.

#### Release notes
<!-- Write a line or two to be included in the release notes.
Everything is worth mentioning, because you did it for a reason. -->

Changed pagination counts in order cycle exchange UI

<!-- Please assign one category to your PR and delete the others. 
The categories are based on https://keepachangelog.com/en/1.0.0/. -->

Changelog Category: Changed

![all_variants](https://user-images.githubusercontent.com/9029026/71523754-a7af2680-28ca-11ea-9d26-a71b57e9e361.png)

luisramos0 commented:
ah, one more point: I think we should change to "39 of 49 **Variants** Loaded" to be more clear.

luisramos0 commented on app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee:
```diff
@@ -51,9 +58,8 @@ angular.module('admin.orderCycles')
-      ExchangeProduct.index params, (products, num_of_pages, num_of_products) ->
+      ExchangeProduct.index params, (products, num_of_pages) ->
         enterprise.num_of_pages = num_of_pages
-        enterprise.num_of_products = num_of_products
```
instead of an extra method above "exchangeLoadedVariants" I'd implement the logic here in loadExchangeProducts.
instead of enterprise.num_of_products = num_of_products
we can do:

initialize enterprise.loaded_variants = 0 before the call to  ExchangeProduct.index
`enterprise.loaded_variants = 0`

and then inside:
```
angular.forEach products, (product) ->
        enterprise.loaded_variants += product.variants.length
```

Matt-Yorkley replied:
I can't tell if there's any real benefit in one over the other? Do you think it would be more efficient to increment the count as the variants are loaded?

luisramos0 replied:
I think it would be slightly faster and less complex with one less method. but I dont think it's a massive difference, that's why I approve the PR as is. it's your call :-)
my first comment in 2020 happy new decade!

mkllnk replied:
Hm, maybe there is not much in it but I imagine Luis' version to be more efficient. The variants are counted once after they have been loaded. In the current version, they are calculated every time `exchangeLoadedVariants` is called. I don't know how often Angular executes that.

And since we are already talking about preferred styles, I have another version here: :wink: 
```coffee
enterprise.loaded_variants = products.reduce((sum, product) => sum + product.variants.length, 0))
```

luisramos0 replied:
:-)

luisramos0 commented:
yesh, thanks @mkllnk 

@RachL shall I change from "Products Loaded" to "Variants Loaded"?

RachL commented:
@luisramos0 yes maybe that's best. Worse case we can change in translations :)

Matt-Yorkley commented:
@mkllnk and @luisramos0 

I've renamed "Products" to "Variants", updated the `loaded_variants` counting methods (second suggestion didn't work), and found and fixed a pre-existing javascript console error related to clashing `$digest` cycles.

luisramos0 commented on app/assets/javascripts/admin/order_cycles/controllers/incoming_controller.js.coffee:
```diff
@@ -28,4 +28,3 @@ angular.module('admin.orderCycles').controller 'AdminOrderCycleIncomingCtrl', ($
   $scope.selectAllVariants = (exchange, selected) ->
     $scope.loadAllExchangeProducts(exchange).then ->
       $scope.setExchangeVariants(exchange, $scope.suppliedVariants(exchange.enterprise_id), selected)
-      $scope.$apply()
```
was this causing the console error? this is needed for some cases... the case is:
- on the incoming exchanges when you have only a few pages loaded and you "select all", it will load all variants first and then select them all. in this case, this apply() is making the "select all" select box become selected, otherwise it wont re-render after all variants are loaded and selected and it will remain unchecked (which is incorrect).

maybe we need to run this apply only in this case, which is the case where loadAllExchangeProducts actually loads something. I'd guess the console error is for the case where loadAllExchangeProducts is not loading anything (all products are already loaded)

Matt-Yorkley replied:
Maybe I'll just revert it and open a tech-debt issue for now :+1:

Matt-Yorkley replied:
Issue: #4676

lin-d-hop commented:
On testing this I noticed a couple things:

1) This doesn't show for single producer shops. There is no count at all. I assume this is out of scope for this issue.

2) A few weird cases testing on UK staging with Cambridge Food Hub
![VariantCount](https://user-images.githubusercontent.com/8669848/72801302-22a4fc00-3c41-11ea-8d06-15e9547d96a8.png)

The values are different in the Tab and in the Panel (when you click to open the tab. 160 vs 163.
Consistently is says x/0 selected. Surely the total should be 163, not 0.

Ping @matt-yorkley.

luisramos0 commented:
Looks like the problem you are seeing is unrelated to this PR...
I dont understand why it only fails here, but this change will fix the problem:

```diff
--- a/app/controllers/api/exchange_products_controller.rb
+++ b/app/controllers/api/exchange_products_controller.rb
@@ -31,7 +31,7 @@ module Api
       render text: {
         count: Spree::Variant.
           not_master.
-          where(product_id: products).
+          where(product_id: products.select(&:id)).
           count
       }.to_json
     end
```

luisramos0 commented:
It's not just this problem, there's something else going on. I copied uk staging data to my computer and I am having a look. This enterprise Cambridge FoodHub looks like a really good test case, Ill investigate further :+1:

luisramos0 commented:
This test case is very good!
I found another bug in the counting of variants. The simple count of variants (to display the summary before any tab is opened) is not considering order_cycle config that sets "coordinator inventory only". For example in Hodmedods MASTER there are 46 inventory items that are hidden and those variants are considered in the counting but are not displayed...

This problem is closely related to this PR but it is pre-existing. But this PR makes it worse because the numbers before this PR were variants against products (not obvious to see the problem). Now it's variants against variants and the issue is obvious!
I think I'll create another PR to fix the problem but we should mark this blocked and only merge this when the new one I'll create is also ready to be merged.

Let me know if you dont understand something I said above.

luisramos0 commented:
Another problem I see is that the pagiantion is done on the products but we display variants.
Each page is 100 products but that is displayed as variants, so we get random page sizes like here:

![Screenshot 2020-01-24 at 10 18 44](https://user-images.githubusercontent.com/1640378/73061891-0ef0d400-3e93-11ea-8dfb-d4ca45d607ee.png)

luisramos0 commented:
ok, #4718 is ready for code review.

luisramos0 commented:
Now that 4718 is merged, we should rebase this and test it again 💪

Matt-Yorkley commented:
Rebased :+1:

lin-d-hop commented:
@Matt-Yorkley  more conflicts :-(

Matt-Yorkley commented:
Conflict resolved :+1:

filipefurtad0 commented:
Hey @Matt-Yorkley,

This looks great. Below - left, before - and - right, after:

![image](https://user-images.githubusercontent.com/49817236/74255406-bcfcda80-4ce9-11ea-8343-7169eac05633.png)

One now sees all the variants.

I think this is ready to go.

@lin-d-hop : You have tested this before, would you like to have a look still?

RachL commented:
I think this is good to go, no need for @lin-d-hop to review it :-)
